### PR TITLE
roachtest: geo distributed roachtests which specify a number of nodes…

### LINKED ISSUE #92150: Geo distributed roachtests which specify a node count not equal to number of default zones causes error.
**Describe the problem**

When running a geo distributed roachtest where node count is > default zone count OR node count < default zone count, a cluster creation error OR index out of bound error are encountered respectively

**To Reproduce**

Run a geo distributed roach test where cluster node count != default zone count (currently 9)

If possible, provide steps to reproduce the behavior:

Run a test with node count  = 4
1. `roachtest run sqlsmith/settup=seed-multi-region/setting=multi-region` 

Cluster creation fails.

Run a test with node count = 10
1. update `roachtest/tests/sqlsmith.go` with node count = 10
2. `roachtest run sqlsmith/settup=seed-multi-region/setting=multi-region` 

Index error

**Expected behavior**
Cluster should be created successfully and test executed.

**Environment:**
 - master
 
**Additional context**
What was the impact? Failed daily roachtests

This is a bug that has existed for quite some time but was only surfaced due to a recent change increasing the number of default zones, hence resulting in an existing test having node count < default zone count.

Jira issue: CRDB-21604

### PR DESCRIPTION
**Pull request by smg260**
… strictly less

than the default number of gcloud zones (currently 9) result in a gcloud syntax error for omission of instance name. This happens because we loop through `len(zones)` instead of `len(nodes)`

Resolves: https://github.com/cockroachdb/cockroach/issues/92150
Release note: none
Epic: none

### CODE REVIEW

**Review by renatolabs — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92140)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @smg260)

___
*[`pkg/roachprod/vm/gce/gcloud.go` line 533 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92140#-NHAoaCD7sood3bXMQf3:-NHAoaCD7sood3bXMQf4:bdc5pvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/88baf159430e46164c98001017dc0f16a8f1ea99/pkg/roachprod/vm/gce/gcloud.go#L533)):*
> ```Go
> 
> 	nodeZones := vm.ZonePlacement(len(zones), len(names))
> 	zoneHostNames := make([][]string, len(names))
> ```

I don't think this fully fixes the issue, as it inverts the assumption: now if we have more nodes than zones, we'll have a similar issue (this time manifested as an index out of bounds panic on `zones[i]` in the loop below). I believe the length here should be `min(numNodes, numZones)`.
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 546 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92140#-NHApfupDTpFtMTXXe7C:-NHApfupDTpFtMTXXe7D:bh8yy0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/88baf159430e46164c98001017dc0f16a8f1ea99/pkg/roachprod/vm/gce/gcloud.go#L546)):*
> ```Go
> 			output, err := cmd.CombinedOutput()
> 			if err != nil {
> 				return errors.Wrapf(err, "Command: gcloud %s\nOutput: %s", args, output)
> ```

As we're touching this file, could we also update the error reporting here? It would be more helpful if this error included `argsWithZone` (the actual arguments passed to `gcloud`) instead of just `args`.

**Review by smg260 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92140)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @renatolabs)

___
*[`pkg/roachprod/vm/gce/gcloud.go` line 533 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92140#-NHAoaCD7sood3bXMQf3:-NHAxA4k9SVIRH9gJBSw:b-a5vlbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/88baf159430e46164c98001017dc0f16a8f1ea99/pkg/roachprod/vm/gce/gcloud.go#L533)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I don't think this fully fixes the issue, as it inverts the assumption: now if we have more nodes than zones, we'll have a similar issue (this time manifested as an index out of bounds panic on `zones[i]` in the loop below). I believe the length here should be `min(numNodes, numZones)`.
</blockquote></details>

Yep thanks - just caught this is running the test :) Was testing your exact suggestion. It's a confusing piece of code not helped by variable naming. 
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 546 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92140#-NHApfupDTpFtMTXXe7C:-NHAzOqN2Cw3FKyM1uc4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/88baf159430e46164c98001017dc0f16a8f1ea99/pkg/roachprod/vm/gce/gcloud.go#L546)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

As we're touching this file, could we also update the error reporting here? It would be more helpful if this error included `argsWithZone` (the actual arguments passed to `gcloud`) instead of just `args`.
</blockquote></details>

Done.

**Review by renatolabs — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92140#-:-NHB9rMw25FDHlqcSmpN:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @herkolategan)

File: pkg/roachprod/vm/gce/gcloud.go [hunk 1]
```diff
@@ -552,6 +552,13 @@ func (p *Provider) Create(
 	return g.Wait()
 }
 
+func min(a, b int) int {
```
Comment (srosenberg) [inline pkg/roachprod/vm/gce/gcloud.go:555]: The duplication of primitive math functions is kinda annoying, thanks to Go's decision to provide only `math.Min(x, y float64)`. 

I was curious to see what other packages have done. There are several variants. E.g., this one does `a <= b` [1] which while isomorphic may behave differently under branch speculations.

I think it's okay to use `math.Min` as long as (a) integer args don't exceed float limits, (b) the code doesn't need to be performant. (Go's runtime will use optimized assembly, but `runtime.convT64` and special cases would add up.)
In any case, it's a nit :)

[1] https://github.com/cockroachdb/cockroach/blob/b0daa5f8d2af1c9c59ec79e09fb74a104c5bcbc2/pkg/sql/opt/props/cardinality.go#L154
[2] https://github.com/golang/go/blob/4f0d3bcd6d2d6bc726044153bf6875d195203b32/src/math/dim_amd64.s#L56

Comment (smg260) [reply pkg/roachprod/vm/gce/gcloud.go:555]: I agree and searched the codebase to see how it was done elsewhere, and ultimately lifted the function.

**Pull request merged**